design/internal-redirect_design.md

```diff
@@ -0,0 +1,182 @@
+
+## Non Goals
+- Supporting user defined 'internal redirect policy' predicate.
+
+## High-Level Design
+
+Envoy Route supports [internal_redirect_policy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/http_connection_management#internal-redirects) and it can be configured [using the API](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routeaction-internal-redirect-policy). The changes proposed in this document will allow the configuration of internal redirect policies in Contour.
+
+At a high level the proposed changes will imply:
+- Adding new fields at virtual host level to configure the internal redirect policy in the YAML.
+- Changing some structs in the code.
+- Generating the corresponding Envoy configuration.
+
+### Proposed YAML fields
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+ name: myservice
+ namespace: prod
+spec:
+ virtualhost:
+   fqdn: www.example.com
+   routes:
+     - conditions:
+         - prefix: /echo
+       services:
+         - name: echo
+           port: 8080
+       internalRedirectPolicy:
+         maxInternalRedirects: 5
+         redirectResponseCodes: [ 301, 302, 303 ]
+         predicates:
+           safeCrossScheme: true
+           previousRoutes: true
+           allowedRouteNames: [ "other" ]
+         allowCrossSchemeRedirect: true
+     - name: other
+       conditions:
+         - prefix: /other
+       services:
+         - name: other
+           port: 8090
+```
+
+## Detailed Design
+
+Add a new type: `HTTPInternalRedirectPolicy` to define when the proxy should intercept and handle redirect responses internally.
+
+```Go
+// HTTPInternalRedirectPolicy defines when to handle redirect responses internally.
+type HTTPInternalRedirectPolicy struct {
+    // MaxInternalRedirects An internal redirect is not handled, unless the number of previous internal redirects that a downstream request has encountered is lower than this value.
+    // +optional
+    MaxInternalRedirects uint32 `json:"maxInternalRedirects,omitempty"`
+
+    // RedirectResponseCodes If unspecified, only 302 will be treated as internal redirect.
+    // Only 301, 302, 303, 307 and 308 are valid values.
+    // +optional
+    RedirectResponseCodes []uint32 `json:"redirectResponseCodes,omitempty"`
+
+    // Predicates are queried when an upstream response is deemed to trigger an internal redirect by all other criteria.
+    // +optional
+    Predicates HTTPInternalRedirectPredicate `json:"predicates,omitempty"`
+
+    // AllowCrossSchemeRedirect Allow internal redirect to follow a target URI with a different scheme than the value of x-forwarded-proto.
+    // +optional
+    AllowCrossSchemeRedirect bool `json:"allowCrossSchemeRedirect,omitempty"`
+}
+
+// HTTPInternalRedirectPredicate defines the predicate used to accept or reject an internal redirection.
+type HTTPInternalRedirectPredicate struct {
```

Comment:
would it make sense to get rid of this additional type and pull these up to the HTTPInternalRedirectPolicy struct? I'm not sure how much value having these nested gives us in this API

we don't necessarily need to use "predicate" in the naming to make things a little clearer for users, I think the Envoy names also may be a little opaque

e.g SafeCrossScheme -> something that conveys "allow mismatched downstream and redirect target scheme" (this one is harder to come up with a good name 😅 ), or PreviousRoutes -> AllowRepeatedRouteRedirect

Reply:
This would also be an opportunity to solve the configuration issue where `SafeCrossScheme` being meaningful only if `AllowCrossSchemeRedirect` redirect is `true`. 

It think both fields should be merged into something like `CrossSchemeRedirect` and take an enum like `always`, `never` and `safeOnly`.

Reply:
I have updated the proposal by removing the nested type, and trying to simplify the API.